ambry-api/src/main/java/com/github/ambry/commons/CallbackUtils.java

```diff
@@ -89,4 +93,56 @@ public static <T> Callback<T> fromCompletableFuture(CompletableFuture<T> future)
       }
     };
   }
-}
+
+  /**
+   * Create a {@link Callback} that invokes the preAction and then the callback with error handling.
+   * It is meant to be a part of multi-step callback chain and handle intermediate failures in a consistent way.
+   * @param <T> the input and result type for the callback.
+   * @param callback the callback to call after the preAction.
+   * @param preAction the optional action to take before the callback is called.
+   * @return the managed {@link Callback}.
+   */
+  public static <T> Callback<T> chainCallback(Callback<T> callback, ThrowingFunction<T, T> preAction) {
```

Comment:
Good point, I think this makes sense. There are 3 places we could decide on whether to introduce metrics
1) In S3BaseHandler, where we do post-processing for all requests like removing Ambry header
2) Individual handler, most are delegated to existing handlers with some pre and post processing
3) The top level of composite handlers like S3HeadHandler and the S3MultipartHandler, so far we probably don't yet have a need since they are just simple delegation.

There is not need for #3 right now. 
I think we should do #2. 
But not sure about #1 ... 

Thoughts?

Reply:
Sure. I also agree with you. We may not need for 3) and 1) yet but may need for 2).

Reply:
Cool, I made the updates.